keps/sig-instrumentation/2580-custom-metrics-router/README.md

```diff
@@ -0,0 +1,658 @@
+which in turn forwards it to the metrics server based on the API group i.e.
+`metrics.k8s.io`, `custom.metrics.k8s.io` or `external.metrics.k8s.io`.
+
+#### API Discovery
+API servers expose endpoints which can be used by a client to find the
+correct resource version and discover information about which API resources
+are provided by the aggregated API server. The _custom-metrics-apiserver_
+similarly exposes an endpoint which lists all the custom metric types(the
+associated resource type and metric name) and external metrics types which
+are provided by the metrics server.
+
+### Solution
+The proposed solution consists of 3 main components. A controller which uses
+a CRD to track metric sources. The controller uses API discovery to find the
+metrics provided by different metrics providers. It then builds a routing
+table mapping the metric to a provider. The second component is the routing
+table. One of the requirements of this routing table is that routing entries
+should be able to be added and deleted without a complete re-indexing of all
+metrics. And finally there is the request router. This component receives the
+metric requests which are proxied from the API server and then consults the
+routing table and queries the correct metric source and returns the result.
+
+![Router Architecture](custom_metrics_router.png)
+
+#### Metrics Discovery Controller
+`CustomMetricsSource` is the proposed CRD which should be used to register
+metrics providers. Both custom and external metrics providers can be tracked
+with this resource.
+
+```golang
+type Service struct {
+    Namespace string
+    Name  	string
+    Port  	int32
+}
+
+type MetricType string
+
+const (
+  CustomMetricsType   = "CustomMetrics"
+  ExternalMetricsType = "ExternalMetrics"
+)
+
+type CustomMetricsSourceSpec struct {
+  Service               Service      `json:"service"`
+  InsecureSkipTLSVerify bool         `json:"insecureSkipTLSVerify"`
+  Priority              int          `json:"priority"`
+  MetricTypes           []MetricType `json:"metricTypes"`
+}
+
+type CustomMetricsSource struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+  Spec              CustomMetricsSourceSpec `json:"spec"`
+}
+```
+
+This CRD uses almost the same structure as the
+[APIService](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#apiservice-v1-apiregistration-k8s-io)
+used to register APIs in the kube-apiserver. In the kube-apiserver higher
+priority `APIservices` for the same API group are always used for queries of
+that type. However when 2 services have the same priority, the requests could
+be sent to either service with equal probability. In the case of the metrics
+service it might not make sense to do that same because the returned metric
+values may differ causing the replica count of the target to fluctuate.
+Instead the age of the metrics API service registration object can be used as
```

Comment:
Could you elaborate on what object age is used here  ? And how ?

Reply:
All kubernetes resources(custom or otherwise) have a creation timestamp was is set when the resource is created. The routing table would contain entries of providers which expose a given metric. This would be sorted in increasing order based on the creation timestamp.

Reply:
My concern with the creation timestamp is that it is not obvious for the user to get it, even displayed in a column it is hard to figure out which one is the most recent among a bunch of resources.

I was wondering if sorting resources by `name` would be a better choice:
* `metadata.name` is an immutable field, if a resource is deleted and then recreated with the same name the creation timestamp will change, but not its name. It makes the behaviour more predictable.
* I don't have a good example for K8S but on OpenShift this is [how SCC are prioritized](https://docs.openshift.com/container-platform/4.7/authentication/managing-security-context-constraints.html#scc-prioritization_configuring-internal-oauth). My experience as a user is that it is a nice way to solve that problem.

What do you think ?

Reply:
The reason for using timestamp is to reduce the chances of accidentally changing the provider which exposes a given metric. This might make it harder to debug but the metric value would be stable. Assume the user registers a new provider and is unaware that it exposes a metric which is currently in use. The new metric value is different to the existing metric value. This means the scaling target would suddenly scale up or down probably causing an outage.
Also to improve debugging I think the routing table could be written out to a configmap which can be examined for debugging. There is also a need to provide overrides for specific metric which this KEP doesn't address but we can cross that bridge when we get there.